### DISCUSSION TITLE
How to pass different configuration to each test?

### DISCUSSION THREAD

**Discussion by szabgab** [rwf2/Rocket#2856] [Questions] [answered]
As I would like each test to be independent from the other tests each one need its own configuration. (e.g. a different database or a different temporary folder).

One way to pass this information is by setting environment variables - but those are per process and thus I had to set the test to run in a single thread.  How can I pass some parameter to each test?

The simplified version of the main file looks like this:

```
#[launch]
fn rocket() -> _ {
    rocket::build()
        .mount(
            "/",
            routes![
                main,
            ],
        )
        .attach(AdHoc::config::<MyConfig>())
}

```

The pseudo version of the test looks like this

```
#[test]
fn test_simple() {
    let rocket_toml_path = some_temporary_place();
    let database_name = random_for_this_test();
    save_rocket_toml with database_name in rocket_toml_path;

    std::env::set_var("ROCKET_CONFIG", rocket_toml_path);

     let client = Client::tracked(super::rocket()).unwrap();
     ...
}
```

The real code is in [this repository](https://github.com/szabgab/meet-os.rs/) in `src/main.rs` and `src/test_simpl.rs`

**Comment by SergioBenitez** [comment, marked as answer]
Using [`Rocket::configure()`](https://api.rocket.rs/v0.5/rocket/struct.Rocket#method.configure):

```rust
#[test]
fn test_simple() {
    let rocket = rocket().configure(provider);
    let client = Client::tracked(rocket);
}
```

`provider` here is any configuration source. If you want it to be a TOML file, then:

```rust
use rocket::figment::providers::{Toml, Format, Env};

let provider = Toml::file("some_file.toml").nested();
```

Maybe you just want to set one parameter, though. In which case, maybe you do something like:

```rust
use rocket::config::Config;

let provider = Config::figment()
    .merge(("databases.foo.url", "some_value"));    
```

**Reply by szabgab** [reply]
Thank you. That's what I was looking for. I wrote a [full example](https://rust.code-maven.com/slides/rust/custom-configuration-and-injecting-values-in-tests.html) in case anyone is interested.
